wardrobe/tradable_kitties/src/lib.rs

```diff
@@ -0,0 +1,389 @@
+    let mut input_coin_data: Vec<DynamicallyTypedData> = Vec::new();
+    let mut output_coin_data: Vec<DynamicallyTypedData> = Vec::new();
+    let input_kitty_to_be_traded: Option<TradableKittyData>;
+
+    let mut total_input_amount: u128 = 0;
+    let total_price_of_kitty: u128;
+
+    if let Ok(td_input_kitty) = input_data[0].extract::<TradableKittyData>() {
+        ensure!(
+            td_input_kitty.price != 0,
+            TradeableKittyError::KittyPriceCantBeZero
+        );
+        input_kitty_to_be_traded = Some(td_input_kitty.clone());
+        total_price_of_kitty = td_input_kitty.price;
+    } else {
+        return Err(TradeableKittyError::BadlyTyped);
+    }
+
+    if let Ok(td_output_kitty) = output_data[0].extract::<TradableKittyData>() {
+        ensure!(
+            input_kitty_to_be_traded.clone().unwrap().kitty_basic_data
+                == td_output_kitty.kitty_basic_data,
+            TradeableKittyError::KittyBasicPropertiesAltered
+        );
+    } else {
+        return Err(TradeableKittyError::BadlyTyped);
+    }
+
+    for coin_data in input_data.iter().skip(1) {
+        let coin = coin_data
+            .clone()
+            .extract::<Coin<ID>>()
+            .map_err(|_| TradeableKittyError::BadlyTyped)?;
+
+        let utxo_value = coin.0;
+        ensure!(
+            utxo_value > 0,
+            TradeableKittyError::MoneyError(MoneyError::ZeroValueCoin)
+        );
+        input_coin_data.push(coin_data.clone());
+        total_input_amount = total_input_amount
+            .checked_add(utxo_value)
+            .ok_or(TradeableKittyError::MoneyError(MoneyError::ValueOverflow))?;
+    }
+
+    for coin_data in output_data.iter().skip(1) {
+        let coin = coin_data
+            .clone()
+            .extract::<Coin<ID>>()
+            .map_err(|_| TradeableKittyError::BadlyTyped)?;
+
+        let utxo_value = coin.0;
+        ensure!(
+            utxo_value > 0,
+            TradeableKittyError::MoneyError(MoneyError::ZeroValueCoin)
+        );
+        ensure!(
+            utxo_value > 0,
+            TradeableKittyError::MoneyError(MoneyError::ZeroValueCoin)
+        );
+        output_coin_data.push(coin_data.clone());
+    }
+    ensure!(
+        total_price_of_kitty <= total_input_amount,
+        TradeableKittyError::InsufficientCollateralToBuyKitty
+    );
+
+    // Filtered coins are sent to MoneyConstraintChecker for money validation.
+    Ok(MoneyConstraintChecker::<0>::Spend.check(&input_coin_data, &[], &output_coin_data)?)
+}
+
+/// Checks if updates to the prices of tradable kitties are possible or not.
+/// Prices of multiple tradable kitties can be updated in the same transaction.
+fn check_kitty_price_update(
+    input_data: &[DynamicallyTypedData],
+    output_data: &[DynamicallyTypedData],
+) -> Result<TransactionPriority, TradeableKittyError> {
+    ensure!(
+        input_data.len() == output_data.len() && !input_data.is_empty(),
```

Comment:
```suggestion
        input_data.len() == output_data.len(),
```
I think this was redundant